doc: Add matrix expressions to docstrings

ykharkov opened the pull request:
Resolves issue https://github.com/amazon-braket/amazon-braket-sdk-python/issues/597

*Description of changes:*
Added definitions of unitary matrices for each gate (docstrings)

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/README.md) and [API docs](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

krneta commented on src/braket/circuits/gates.py:
```diff
@@ -109,6 +109,12 @@ def h(
         Returns:
             Iterable[Instruction]: `Iterable` of H instructions.
 
+        Unitary matrix:
```
I think this might work if you put it below the examples section.

ykharkov replied:
Thanks @krneta. I decided to move it to the top instead, so that it shows the matrix immediately.

math411 commented on src/braket/circuits/gates.py:
```diff
@@ -938,7 +1026,13 @@ def ry(
+    r"""Z-axis rotation gate.
+
+    Unitary matrix:
+        .. math:: \mathtt{R_z}(\phi) = \begin{bmatrix}
```
Do we want to normalize the rotation gates to have theta as their represented angles?

ykharkov replied:
@math411 Suggest to use either phi or theta throughout?
Of course this is a nit.

ykharkov replied:
Renamed theta->phi whenever possible to unify notations

rmshaffer commented on src/braket/circuits/gates.py:
```diff
@@ -1489,7 +1633,15 @@ def xy(
+    r"""Controlled phase shift gate.
+
+    Unitary matrix:
+        .. math:: \mathtt{CPhaseShift00}(\phi) = \begin{bmatrix}
```
```suggestion
        .. math:: \mathtt{CPhaseShift}(\phi) = \begin{bmatrix}
```

rmshaffer commented on src/braket/circuits/gates.py:
```diff
@@ -1627,7 +1787,15 @@ def cphaseshift00(
+
+    Unitary matrix:
+        .. math:: \mathtt{CPhaseShift01}(\phi) = \begin{bmatrix}
+                e^{i \phi} & 0 & 0 & 0 \\
+                0 & 1 & 0 & 0 \\
```
should the `e^{i \phi}` be in the second row here?
```suggestion
                1 & 0 & 0 & 0 \\
                0 & e^{i \phi} & 0 & 0 \\
```

rmshaffer commented on src/braket/circuits/gates.py:
```diff
@@ -2610,12 +2884,25 @@ def gpi2(
 
 class MS(TripleAngledGate):
-    """IonQ Mølmer-Sørenson gate.
+    r"""IonQ Mølmer-Sørenson gate.
```
```suggestion
    r"""IonQ Mølmer-Sørensen gate.
```

rmshaffer commented on src/braket/circuits/gates.py:
```diff
@@ -60,7 +60,13 @@
-    """Hadamard gate."""
+    r"""Hadamard gate.
+
+    Unitary matrix:
+        .. math:: \mathtt{H} = \frac{1}{\sqrt{2}} \begin{bmatrix}
+                1 & 1 \\
+                1 & -1 \end{bmatrix}.
+    """
```
[super nit] the rendering of "Unitary matrix:" is in some weird heading style ([link to rendered doc](https://amazon-braket-sdk-python--756.org.readthedocs.build/en/756/_apidoc/braket.circuits.gate.html#braket.circuits.gate.Gate.H), see screenshot below). I wonder if it can show up as plain text instead, maybe by adjusting whitespace in the docstring somehow?
![image](https://github.com/amazon-braket/amazon-braket-sdk-python/assets/3620100/c66331a0-a8ac-40fe-97df-1920e874a2a0)

math411 replied:
```suggestion
    Unitary matrix:
    
        .. math:: \mathtt{H} = \frac{1}{\sqrt{2}} \begin{bmatrix}
                1 & 1 \\
                1 & -1 \end{bmatrix}.
    """
```

This styling can be removed by adding a space between the blocks of text as with the example above.

ykharkov replied:
Would this be a better option?

![image](https://github.com/amazon-braket/amazon-braket-sdk-python/assets/112575584/14ac5ae5-710c-4486-bab3-fb948dbe01e2)

I don't think we have much control over the formatting here.

rmshaffer replied:
Does @math411's suggestion work?

ykharkov replied:
@rmshaffer @math411 The comments are addressed. This is the updated view.

<img width="842" alt="Screenshot 2023-10-28 at 2 48 05 PM" src="https://github.com/amazon-braket/amazon-braket-sdk-python/assets/112575584/854d6933-9f31-41a7-bf85-540359974dbf">

rmshaffer replied:
awesome!

rmshaffer reviewed (commented):
This is looking awesome. For reference, here is an example of the rendered doc for this PR:
https://amazon-braket-sdk-python--756.org.readthedocs.build/en/756/_apidoc/braket.circuits.gates.html#braket.circuits.gates.CPhaseShift00

General question, do you have any idea how hard it would be to have the same docstring appear in the `circuit` documentation for these gates?
https://amazon-braket-sdk-python--756.org.readthedocs.build/en/756/_apidoc/braket.circuits.circuit.html#braket.circuits.circuit.Circuit.cphaseshift00
or maybe somehow have the `circuit` API doc link back to the `gates` API doc? I'm not sure which page users are more likely to land on, but it would be nice if both lead them to this matrix definition somehow.

math411 commented:
To @rmshaffer's point, we can add this within the subroutine [docstrings](https://github.com/amazon-braket/amazon-braket-sdk-python/pull/756/files#diff-85b1f0c3121c7b89a6f7951f13b6ef494dca0216445ef035fe69ed30e054fdc2R93). This should render on the circuit page but this is something to double check with a single gate first.

rmshaffer commented on src/braket/circuits/gates.py:
```diff
@@ -2610,12 +2919,26 @@ def gpi2(
-    """IonQ Mølmer-Sørenson gate.
+    r"""IonQ Mølmer-Sørensen gate.
+
+        Unitary matrix:
```
nit, to make this indentation consistent with the rest of the gates:
```suggestion
    Unitary matrix:
```

ykharkov commented:
@math411 
I think both me and @rmshaffer did not fully understand your suggestion regarding this comment.
> General question, do you have any idea how hard it would be to have the same docstring appear in the `circuit` documentation for these gates?

Are we OK merging as it is? Otherwise please advise on the next steps.

ajberdy commented:
We will remove the generic line about "registering the subroutine" from the `.h` etc methods of the gates and instead insert the class documentation (with the gate name, matrix etc)

math411 commented:
> @math411 I think both me and @rmshaffer did not fully understand your suggestion regarding this comment:
> 
> > General question, do you have any idea how hard it would be to have the same docstring appear in the `circuit` documentation for these gates?
> 
> > we can add this within the subroutine [docstrings](https://github.com/amazon-braket/amazon-braket-sdk-python/pull/756/files#diff-85b1f0c3121c7b89a6f7951f13b6ef494dca0216445ef035fe69ed30e054fdc2R93). This should render on the circuit page but this is something to double check with a single gate first.
> 
> Are we OK merging as it is? Otherwise please advise on the next steps.

@ykharkov my idea was, as an example, change [L101](https://github.com/amazon-braket/amazon-braket-sdk-python/pull/756/files#diff-85b1f0c3121c7b89a6f7951f13b6ef494dca0216445ef035fe69ed30e054fdc2R101) to:

```
        r"""Registers this function into the circuit class.
        
        Unitary matrix:

                .. math:: \mathtt{H} = \frac{1}{\sqrt{2}} \begin{bmatrix}
                        1 & 1 \\
                        1 & -1 \end{bmatrix}.
    """
```

rmshaffer commented:
> We will remove the generic line about "registering the subroutine" from the `.h` etc methods of the gates and instead insert the class documentation (with the gate name, matrix etc)

@ajberdy could you clarify the "we" you are referring to? Should the change be made as part of this PR, or is that going to happen separately?

ajberdy commented:
> > We will remove the generic line about "registering the subroutine" from the `.h` etc methods of the gates and instead insert the class documentation (with the gate name, matrix etc)
> 
> @ajberdy could you clarify the "we" you are referring to? Should the change you're suggesting be made as part of this PR, or is that going to happen separately?

@math411 has updated the PR to include the matrices in the circuit subroutine method. I recommend the PR is updated to replace "Registers this function into the circuit class." with the description of the gate for the class doc string, i.e. "Pauli-X gate."

math411 commented:
Added in https://github.com/amazon-braket/amazon-braket-sdk-python/pull/800 to have a more programmatic way of doing this.

ykharkov commented:
> Added in #800 to have a more programmatic way of doing this.
> 
> Accidentally pushed those changes to this branch, I can roll back those changes in favor of this approach (or a more clean version of this).

Thanks @math411 and @ajberdy .

I addressed concern of @ajberdy about "Registers this function into the circuit class." line and replaced with an appropriate gate description.

Let's merge this PR to keep the ball moving and in the followup PR from @math411 we can add a more programatic solution.

Thanks

ajberdy reviewed (approved):
Capturing that we will have to remove the "Args" section from the class doc strings as that is actually broken right now ([doc link](https://amazon-braket-sdk-python.readthedocs.io/en/latest/_apidoc/braket.circuits.gates.html#braket.circuits.gates.Rx)). Though it is broken before and after this PR, so not going to block.
